src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -120,6 +124,62 @@ protected override void OnEventWritten(EventWrittenEventArgs eventData)
     private static Dictionary<string, string> GetEnableEventsArguments(EventCountersInstrumentationOptions options) =>
         new() { { "EventCounterIntervalSec", options.RefreshIntervalSecs.ToString() } };
 
+    /// <summary>
+    /// If instrument name is too long, abbreviates event source name.
+    /// E.g. instrument for `Microsoft-AspNetCore-Server-Kestrel`, `tls-handshakes-per-second`
+    /// would be too long (64 chars), so it's abbreviated to `ec.m.a.s.k.tls-handshakes-per-second`
+    /// instead of `ec.Microsoft-AspNetCore-Server-Kestrel.tls-handshakes-per-second`.
+    ///
+    /// If after that instrument name is still invalid, it will be validated and ignored later in the pipeline.
+    /// </summary>
+    private static string GetInstrumentName(string sourceName, string eventName)
+    {
+        int totalLength = Prefix.Length + 1 + sourceName.Length + 1 + eventName.Length;
+        if (totalLength <= MaxInstrumentNameLength)
+        {
+            return string.Concat(Prefix, ".", sourceName, ".", eventName);
+        }
+
+        int maxEventSourceNameLength = MaxInstrumentNameLength - Prefix.Length - 2 - eventName.Length;
+        var abbreviation = new StringBuilder(maxEventSourceNameLength);
+        int ind = 0;
+        while (ind >= 0 && ind < sourceName.Length)
+        {
+            while (ind < sourceName.Length && (sourceName[ind] == '.' || sourceName[ind] == '-'))
+            {
+                ind++;
+            }
+
+            if (ind < sourceName.Length)
```

Comment:
Check for `abbreviation.Length < maxEventSourceNameLength` as well.

Reply:
sure, but what should we do if it's longer? Currently, it's suppressed by the metrics pipeline in SDK because of the name length in the same way as before.

Reply:
We could either let the Metrics SDK drop it or we could prioritize on Event name. That would mean we first count the number of characters for prefix and eventName and the two `.` characters. If this already exceeds 63 then we don't do anything about it and let the SDK drop it. If not, we try to squeeze in as much of the source name as we can. That might lead to omission of some of the initial characters of the source name and might make the name unintuitive. But we would still have the name suffixed with the full event name and that's probably better than dropping it entirely because the source name is long.

Having said that, this is already a step in the right direction. We could always extend functionality in follow-up PRs.

Reply:
Makes sense! I implemented it, but I preserve the initial characters in event source name and drop the rest once it's impossible to squeeze more of them. PTAL

Reply:
We are still not utilizing the full 63 characters length limit.

Consider this example:

SourceName: "CustomEventSource" (17 characters)
EventName: "longEventName-abcdefghijklmnopqrstuvwxyz-1234567890" (51 characters)

What should we return as the instrument name for this example?

With the current changes in the PR, we would return:

"ec.**c**.longEventName-abcdefghijklmnopqrstuvwxyz-1234567890" (56 characters)

While we had room for 7 more characters which we didn't use at all. In that case, the result would have been:
"ec.**CustomEv**.longEventName-abcdefghijklmnopqrstuvwxyz-1234567890" (63 characters)

I guess we should try to add as many characters as we can, but I would like to know your thoughts.